libs/wingsdk/src/simulator/lockfile.ts

```diff
@@ -0,0 +1,125 @@
+import type { PathLike } from "fs";
+import fs, { type FileHandle } from "fs/promises";
+import path from "path";
+import { isNodeError } from "./util.js";
+
+// eslint-disable-next-line @typescript-eslint/no-shadow
+const exists = async (path: PathLike) => {
+  try {
+    await fs.stat(path);
+    return true;
+  } catch {
+    return false;
+  }
+};
+
+const UPDATE_INTERVAL = 1000;
+const STALE_THRESHOLD = 5000;
+
+export interface LockfileProps {
+  readonly path: string;
+  readonly onCompromised?: () => void;
+}
+
+export class Lockfile {
+  private path: string;
+  private lockfile: FileHandle | undefined;
+  private timeout: NodeJS.Timeout | undefined;
+  private lastMtime: number | undefined;
+  private onCompromised: ((reason: string) => void) | undefined;
+
+  public constructor(props: LockfileProps) {
+    this.path = props.path;
+    this.onCompromised = props.onCompromised;
+  }
+
+  public async lock() {
+    if (this.lockfile) {
+      return; // already locked
+    }
+
+    await fs.mkdir(path.dirname(this.path), { recursive: true });
+
+    // If lockfile exists but it's stale, remove it.
+    if (await exists(this.path)) {
+      try {
+        const stats = await fs.stat(this.path);
+        if (Date.now() - stats.mtimeMs > STALE_THRESHOLD) {
+          await fs.rm(this.path);
+        }
+      } catch (error) {
+        throw new Error("Failed to check lockfile status: " + error);
+      }
+    }
+
+    try {
+      this.lockfile = await fs.open(this.path, "wx");
+    } catch (error) {
+      if (isNodeError(error) && error.code === "EEXIST") {
+        throw new Error(
+          "Another instance of the simulator is already running. Please stop the current simulation before starting a new one."
+        );
+      }
+      throw error;
+    }
+
+    const updateLockfileUtimes = () => {
+      this.timeout = setTimeout(() => {
+        void (async () => {
+          // Check if the lockfile got compromised because we were too late to update it.
+          if (this.lastMtime && Date.now() > this.lastMtime + STALE_THRESHOLD) {
+            this.markAsCompromised("Lockfile was not updated in time");
+            return;
+          }
+
+          // Check if the lockfile got compromised because access was lost or something else updated it.
+          if (this.lastMtime) {
+            try {
+              const stats = await fs.stat(this.path);
+              if (stats.mtimeMs !== this.lastMtime) {
+                this.markAsCompromised(
+                  "Lockfile was updated by another process"
+                );
+                return;
+              }
+            } catch (_error) {
```

Comment:
Maybe it's PTSD but I get weary whenever we're dropping error information since the only other way for a user to get it would be to change the source code and rebuild the project. Can we log the original error some way? (perhaps only if DEBUG=1 or something...)

Reply:
👍🏻 I believe there's already a log mechanism in place, will have a look.

Reply:
Now, if there's a related error, it's passed to the onCompromised callback so the simulator can log it.